### PR TITLE
Overlay Support

### LINKED ISSUE #88: Overlay Network Support
Currently it is required to expose ports in order for Interlock to find the node that the container is using.  This would add overlay network support to "attach" to the container network and use the overlay instead of requiring a globally published port.

### PR DESCRIPTION
**Pull request by ehazlett**
This adds overlay networking support to Interlock meaning you no longer need to publish ports.

This is a significant refactor for the extensions to allow for better use and more reliable event handling.  It also updates to no longer require a volume for the proxy containers as that was unreliable in specific configurations.  This places the configs directly in the container on update to prevent inconsistency with the volumes.

Closes #88

### CODE REVIEW

**Comment by ehazlett** [comment]
There is some issue with Travis that isn't using the Godep properly.  Tests pass locally.  A test image is also available at `ehazlett/interlock:overlay`.

File: docs/extensions.md [hunk 1]
```diff
@@ -10,10 +10,10 @@ Docker images from the Docker Hub for each extension.
 # Load Balancing
 The following load balancing extensions are supported:
 
-## HAProxy
-[HAProxy](http://www.haproxy.org/) is a high performance TCP/HTTP load balancer.
-It is recommended to use the official [Docker Hub HAProxy Image](https://hub.docker.com/_/haproxy/).
+- [HAProxy](./extensions/haproxy.md)
+- [Nginx](./extensions/nginx.md)
 
+<<<<<<< HEAD
```
Comment (everett-toews) [inline docs/extensions.md]: Merge artifact here.

File: ext/lb/haproxy/generate.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+		// check for networking
+		if n, ok := utils.OverlayEnabled(cInfo.Config); ok {
+			log().Debugf("configuring docker network: name=%s", n)
+
+			// FIXME: for some reason the request from dockerclient
+			// is not returning a populated Networks object
+			// so we hack this by inspecting the Network
+			// we should switch to engine-api/client -- hopefully
+			// that will fix
+			//net, found := cInfo.NetworkSettings.Networks[n]
```
Comment (everett-toews) [inline ext/lb/haproxy/generate.go]: If there was a populated Networks object, would you even need the `interlock.network` label on a container?

You could just a populated Networks object to discover the network(s) of a container and the associated address(es).

It seems redundant to have to specify the network twice when running containers like

```
docker run --detach \
  --expose 8080 \
  --net mynetwork \
  --label interlock.network=mynetwork \
  --label interlock.hostname=test \
  --label interlock.domain=local \
  ehazlett/docker-demo
```

I suppose it's a moot point samalba/dockerclient gets replaced.

Comment (ehazlett) [inline ext/lb/haproxy/generate.go]: I agree however you could use both networking modes which is why I require it to be specified.  For example, you might expose a service such as nginx on a public NAT port but access an internal service such as a database.  Therefore, I didn't want to assume.  Perhaps in a future version we could default to the internal but I didn't want such a breaking change at this point.

**Pull request merged**